include/wimlib/compiler.h

```diff
@@ -142,21 +157,127 @@
 #endif
 
 /* STATIC_ASSERT() - verify the truth of an expression at compilation time.  */
+#ifdef _MSC_VER
+#define STATIC_ASSERT(expr) assert(expr)
+#else
 #ifdef __CHECKER__
 #  define STATIC_ASSERT(expr)
 #elif __STDC_VERSION__ >= 201112L
 #  define STATIC_ASSERT(expr)	_Static_assert((expr), "")
 #else
 #  define STATIC_ASSERT(expr)	((void)sizeof(char[1 - 2 * !(expr)]))
 #endif
-
+#endif
 /* STATIC_ASSERT_ZERO() - verify the truth of an expression at compilation time
  * and also produce a result of value '0' to be used in constant expressions */
+#ifdef _MSC_VER
+#pragma comment(lib, "ntdll")
+#define STATIC_ASSERT_ZERO(expr) 0
+#else
 #define STATIC_ASSERT_ZERO(expr) ((int)sizeof(char[-!(expr)]))
-
+#endif
 #define CONCAT_IMPL(s1, s2)	s1##s2
 
 /* CONCAT() - concatenate two tokens at preprocessing time.  */
 #define CONCAT(s1, s2)		CONCAT_IMPL(s1, s2)
+#ifndef PACKAGE_VERSION
+#define PACKAGE_VERSION ""
+#endif
+#ifndef PACKAGE_BUGREPORT
+#define PACKAGE_BUGREPORT ""
+#endif
+#ifdef _MSC_VER
+//We are using Microsoft's MSVC, gcc specific functions are not available
+#include<assert.h>
+#define __attribute__(x)
+#define POINTER_FIX() (size_t)
+#define smart_array(type, name, size) type * name=_alloca((size)*(sizeof(type)))
+#define restrict
+#define EMPTY 0
+#define FILE_SHARE_VALID_FLAGS FILE_SHARE_DELETE | FILE_SHARE_READ |FILE_SHARE_WRITE
+#include<wchar.h>
+static inline wchar_t * wmempcpy(wchar_t *_S1, wchar_t const *_S2, size_t _N)
+{
+	return wmemcpy(_S1,_S2,_N)+_N;
+
+}
+#define alloca _alloca
+
+#include <intrin.h>
+#include <stdint.h>
+uint32_t __inline __builtin_ctz(uint32_t value)
+{
+	unsigned long trailing_zero = 0;
+	if (_BitScanForward(&trailing_zero, value))
+		return trailing_zero;
+	return 32;
+}
 
-#endif /* _WIMLIB_COMPILER_H */
+uint32_t __inline __builtin_clz(uint32_t value)
+{
+	unsigned long leading_zero = 0;
+	if (_BitScanReverse(&leading_zero, value))
+		return 31 - leading_zero;
+	return 32;
+}
+#define __builtin_constant_p(x)	 0
+#define __builtin_prefetch(x, y) 0
```

Comment:
Prefetch is available for MSVC on x86 (but not on ARM) so you should be able to use:
```
#if defined(_M_IX86) || defined(_M_X64)
#define prefetchr		_m_prefetch
#define prefetchw		_m_prefetchw
#else
#define prefetchr(x)
#define prefetchw(x)
#endif
```